### ISSUE TITLE
Understanding Yolo

### ISSUE THREAD

**Issue by GaelGil** [GaelGil/notebooks#37]
I've been reading the yolo v1 paper [here](https://pjreddie.com/media/files/papers/yolo_1.pdf) and I am having trouble understanding it. 
From my understanding so far Yolo creates a grid over the image. Each grid predicts a certain amount of bounding boxes (the paper says by default its 2). Each bounding box is given by x,y,w,h. Along with a confidence score that there is an object there. If the center of an object is in a certain grid cell then that grid cell will detect that object. Each grid cell also predicts the class of the object in the cell. The structure of the network is 24 convolutional layers and 2 fully connected layers. The convolutional layers are used to get features from the image. We only check the image once.

I am having trouble understanding what happens as the image moves forward in the network?
For example in a cnn we extract features from the image then pass that through a fully connected network to get our prediction. 

But what happens in yolo. From my understanding we pass each grid cell through the network getting our predictions class, x, y, w, h, confidence. 
If so how do we get to one bounding box covering the whole object?

**Comment by camoverride** [comment]
There are multiple "levels" at which you can understand the problem.

One level, which is based on the _structure of the network_, is a bit tricky and I'm honestly not super qualified to talk about how it works because I'm not familiar with YOLO - this level of understanding would inform you about how the network specifics choose certain features and pass them through the network.

Another level, which is a bit more abstract but also more practical, is to understand what the goal of the learning is. YOLO reframes object recognition as a regression problem. Regression, how? Well, consider what the cost function is: YOLO predicts some bounding boxes, and compares them to the actual bounding box (in the data set, a human drew boxes around the objects in the image). The loss is computed as the [Jaccard index](https://giou.stanford.edu/) aka Intersection over Union. If the predicted box perfectly overlaps the label, the loss is 0 -- if the predicted box doesn't intersect the label at all, then loss is high.

I don't know specifically how the network does feature selection, but understanding the loss function should give you some hints: it's probably trying to learn the _x, y, w, h_ coordinates spatially, which will involve dropout, max pooling etc. This is "higher level" learning and takes place _after_ generic object representation - I believe that YOLO uses initial layers that were pretrained on imagenet, then adds its own layers on top followed by a fully connected final layer. For image recognition, you often borrow the initial layers of your network from a generic system like imagenet - you can imagine this as "pre-processing" the image into a representation that's generically useful for downstream image recognition tasks. You then add your own layers on top. For some kinds of problems, all it takes is a single fully-connected layer on top of image-net.
